java/arcs/core/storage/api/Handle.kt

```diff
@@ -11,23 +11,28 @@
 
 package arcs.core.storage.api
 
+import arcs.core.data.HandleMode
+
 /** Base interface for all handle classes. */
 interface Handle {
     val name: String
+
+    /** Indicates whether this handle is read-only, write-only, or read-write. */
+    val mode: HandleMode
+
+    /** Assign a callback when the handle is synced. */
+    fun onSync(action: () -> Unit)
```

Comment:
I could see some circumstances where it might be useful to pass the handle to the action, like if you were to share one listener amongst all handles.  But I suppose its ok for our main lifecycle usecase: 

```
val synced = mutableSetOf()
handles.forEach { handle ->
   handle.onSync {
       synced.add(handle)
       checkAll()
   }
   handle.onDesync { 
       synced.remove(handle) 
       checkAll()
  }
} 
fun checkAll() {
   if (handles.size == synced.size) particle.onReady()
}
```

This creates O(# handles * 2) lambda allocations tho.

Reply:
I thought we could just define new lambdas, but you're right, that might be a little inefficient. I've added the callback param.